False positive: const_static_lifetime

<!--
Hi there! Whether you've come to make a suggestion for a new lint, an improvement to an existing lint or to report a bug or a false positive in Clippy, you've come to the right place.

If you want to report that Clippy does not compile, please be sure to be using the *latest version* of *Rust nightly*! Compiler plugins are highly unstable and will only work with a nightly Rust for now. If you are but still have a problem, please let us now!

Thank you for using Clippy!

Write your comment below this line: -->

```sh
$ rustc --version
rustc 1.25.0-nightly (0c6091fbd 2018-02-04)
$ cargo clippy --version
0.0.186
```

Running `cargo clippy` over my project with `#![cfg_attr(feature = "cargo-clippy", deny(clippy))]` set results in the following errors:

```text
error: Constants have by default a `'static` lifetime

    const KEYWORD: &'static str = "whatever";
                   -^^^^^^^----- help: consider removing `'static`: `&str`
```

Doing so promptly results in

```text
error[E0106]: missing lifetime specifier

    const KEYWORD: &str = "whatever";
                   ^ expected lifetime parameter
```

I presume that following Clippy's advice and running into a compiler-enforced hard error is a bug in Clippy?

In this particular case, the const in question is a trait associated constant:

```rust
trait Keyword {
    const KEYWORD: &'static str;
}
```

and the positions it is reporting for change are implementations of it:

```rust
impl Keyword for Whatever {
    const KEYWORD: &'static str = "whatever";
}
```

Removing any `'static` marker, whether in the trait declaration or in implementations of it, causes E0106 to occur.

--- a/clippy_lints/src/const_static_lifetime.rs
+++ b/clippy_lints/src/const_static_lifetime.rs
@@ -87,21 +87,5 @@ impl EarlyLintPass for StaticConst {
         }
     }
 
-    fn check_trait_item(&mut self, cx: &EarlyContext, item: &TraitItem) {
-        if !in_macro(item.span) {
-            // Match only constants...
-            if let TraitItemKind::Const(ref var_type, _) = item.node {
-                self.visit_type(var_type, cx);
-            }
-        }
-    }
-
-    fn check_impl_item(&mut self, cx: &EarlyContext, item: &ImplItem) {
-        if !in_macro(item.span) {
-            // Match only constants...
-            if let ImplItemKind::Const(ref var_type, _) = item.node {
-                self.visit_type(var_type, cx);
-            }
-        }
-    }
+    // Don't check associated consts because `'static` cannot be elided on those (issue #2438)
 }

--- a/tests/ui/const_static_lifetime.stderr
+++ b/tests/ui/const_static_lifetime.stderr
@@ -78,23 +78,5 @@ error: Constants have by default a `'static` lifetime
 24 | const VAR_ARRAY: &'static [u8; 1] = b"T"; // ERROR Consider removing 'static.
    |                  -^^^^^^^-------- help: consider removing `'static`: `&[u8; 1]`
 
-error: Constants have by default a `'static` lifetime
-  --> $DIR/const_static_lifetime.rs:40:23
-   |
-40 |     const TRAIT_VAR: &'static str;
-   |                      -^^^^^^^---- help: consider removing `'static`: `&str`
-
-error: Constants have by default a `'static` lifetime
-  --> $DIR/const_static_lifetime.rs:44:22
-   |
-44 |     const IMPL_VAR: &'static str = "var";
-   |                     -^^^^^^^---- help: consider removing `'static`: `&str`
-
-error: Constants have by default a `'static` lifetime
-  --> $DIR/const_static_lifetime.rs:48:23
-   |
-48 |     const TRAIT_VAR: &'static str = "foo";
-   |                      -^^^^^^^---- help: consider removing `'static`: `&str`
-
-error: aborting due to 16 previous errors
+error: aborting due to 13 previous errors
 

